pkg/controller/v1alpha2/experiment/experiment_controller.go

```diff
@@ -91,6 +109,113 @@ func (r *ReconcileExperiment) Reconcile(request reconcile.Request) (reconcile.Re
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
+	original := instance.DeepCopy()
+
+	if instance.IsCompleted() {
+
+		return reconcile.Result{}, nil
+
+	}
+	if !instance.IsCreated() {
+		//Experiment not created in DB
+		err = util.CreateExperimentinDB(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if instance.Status.StartTime == nil {
+			now := metav1.Now()
+			instance.Status.StartTime = &now
+		}
+		msg := "Experiment is created"
+		instance.MarkExperimentStatusCreated(util.ExperimentCreatedReason, msg)
+		requeue = true
+	} else {
+		// Experiment already created in DB
+		err := r.ReconcileExperiment(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(original.Status, instance.Status) {
+		//assuming that only status change
+		err = util.UpdateExperimentStatusinDB(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		err = r.Status().Update(context.TODO(), instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: requeue}, nil
+}
+
+func (r *ReconcileExperiment) ReconcileExperiment(instance *experimentsv1alpha2.Experiment) error {
+	var err error
+	trials := &trialsv1alpha2.TrialList{}
+	labels := map[string]string{"experiment": instance.Name}
+	lo := &client.ListOptions{}
+	lo.MatchingLabels(labels).InNamespace(instance.Namespace)
+
+	err = r.List(context.TODO(), lo, trials)
+	if err != nil {
+		return err
+	}
+	util.UpdateExperimentStatus(instance, trials)
+
+	reconcileRequired := !instance.IsCompleted()
+	if err != nil {
+		return err
+	}
+	if reconcileRequired {
+		r.ReconcileTrials(instance)
+	}
+	return err
+}
+
+func (r *ReconcileExperiment) ReconcileTrials(instance *experimentsv1alpha2.Experiment) error {
+	var err error
+	parallelCount := 0
+
+	if instance.Spec.ParallelTrialCount != nil {
+		parallelCount = *instance.Spec.ParallelTrialCount
+	} else {
+		parallelCount = 3
+	}
+	activeCount := instance.Status.TrialsRunning
+	succeededCount := instance.Status.TrialsSucceeded
+
+	if activeCount > parallelCount {
+		deleteCount := activeCount - parallelCount
+		if deleteCount > 0 {
+			//delete 'deleteCount' number of trails. Sort them?
```

Comment:
when `activeCount > parallelCount`, I think just keep it as it is.
wonder how to step into `activeCount > parallelCount`. maybe we should control to avoid this state

Reply:
It can happen as there Get and List can return stale data